index.html

```
@@ -174,95 +180,122 @@
       </div>
       <section>
         <h2>
-          Syntax
+          Validation
         </h2>
         <p>
-          When URLs are used for payment method identifiers they MUST be
-          absolute URLs including, at most, a <a>scheme</a>, <a>host</a>,
-          <a>port</a> and <a>path</a>. The URL must be a <a>potentially
-          trustworthy URL</a> as defined in the [[SECURE-CONTEXTS]]
-          specification. Identifier URLs MUST have null <a>query</a> and
-          <a>fragment</a> values.
+          The steps to <dfn data-lt="validate|valid">validate a URL-based
+          payment method identifier</dfn> are given by the following algorithm.
+          The algorithm takes a <var>input</var> string as input and returns
+          either true or an error:
         </p>
+        <ol class="algorithm">
```

This algorithm would be better operating on parsed URLs, not on strings. Most places that call it seem to serialize the parsed URL before calling it, which is pretty pointless since it's just going to parse it again.

---

This is currently not the case in the Payment Request spec. In the "Process payment methods:", I was going to call "validate" just before step 4.2.2:

> Let serializedData be the result of JSON-serializing paymentMethod.data into a string, if the data member of paymentMethod is present, or null if it is not. Rethrow any exceptions.

I can call the URL Parser in the Payment Request spec (re-throwing any errors), and then pass the URL into validate.

---

I can try to take another look at the callers tomorrow. What I remember is that all the callers in this document had to go through parse+serialize+parse. But maybe taking URLs as input is better anyway, not sure yet.

---

OK, after removing the matching algorithm, I see your perspective. But, I think that it's still a better algorithm if it's more "strongly typed", i.e. operates on URLs.

I have a separate issue which is that I think it should return true or false, instead of returning true or throwing a TypeError. E.g. in the payment-method-manifest spec we aren't in a context where throwing makes sense, so we'd have to "catch" the exception, even though we're not even in a JavaScript context where exceptions are sensible to talk about.

---

Agree. I'll make the change to only return true/false and move the exception handling to the PR spec.